src/librustc_passes/ast_validation.rs

```diff
@@ -110,6 +110,31 @@ impl<'a> AstValidator<'a> {
         }
     }
 
+    fn no_negatives_in_bounds(&self, bounds: &GenericBounds) {
+        for bound in bounds {
+            if let GenericBound::Trait(ref poly, TraitBoundModifier::ErroneousNegative) = *bound {
+                let sp = poly.span;
+                let mut err = self.err_handler().struct_span_err(sp,
+                    "negative trait bounds are not supported");
+                if let Ok(prev_source) = self.session.source_map().span_to_prev_source(sp) {
+                    // Find the preceding `+` or `:`
+                    if let Some(spaces) = prev_source.rsplitn(2, &['+', ':'] as &[_]).next() {
+                        let span_to_remove = sp.with_lo(
+                            BytePos(sp.lo().0 - 1 - spaces.len() as u32));
+                        err.span_suggestion_with_applicability(span_to_remove,
+                            "remove this trait bound",
+                            "".into(), Applicability::MachineApplicable);
```

Comment:
This is still too crude to be machine-applicable:

```rust
trait Tr3: !Sync + Copy {}
```

will be left as broken `Tr3 + Copy {}`. Does anyone have a suggestion? I will see if moving this back to the parser provides more context to fix such cases properly.